src/frame.rs

```diff
@@ -84,75 +83,79 @@ impl Frame {
 					}
 					params
 				},
-			});
-		};
+			};
 
-		if frame.has_preference() {
-			fields_count += 1;
-			ret = Some(Self::Preference {
+			// return statement
+			Ok(ret)
+		} else if frame.has_preference() {
+			// return preference
+			Ok(Self::Preference {
 				preference: frame.take_preference(),
-			});
-		};
-
-		if frame.has_attachment() {
-			fields_count += 1;
+			})
+		} else if frame.has_attachment() {
 			let attachment = frame.take_attachment();
-			ret = Some(Self::Attachment {
+
+			// return attachment
+			Ok(Self::Attachment {
 				data_length: attachment.get_length().try_into().unwrap(),
 				id: attachment.get_attachmentId(),
 				row: attachment.get_rowId(),
 				data: None,
-			});
-		};
-
-		if frame.has_version() {
-			fields_count += 1;
-			ret = Some(Self::Version {
+			})
+		} else if frame.has_version() {
+			// return version
+			Ok(Self::Version {
 				version: frame.get_version().get_version(),
-			});
-		};
-
-		if frame.has_end() {
-			fields_count += 1;
-			ret = Some(Self::End);
-		};
-
-		if frame.has_avatar() {
-			fields_count += 1;
+			})
+		} else if frame.has_end() {
+			Ok(Self::End)
+		} else if frame.has_avatar() {
 			let mut avatar = frame.take_avatar();
-			ret = Some(Self::Avatar {
+
+			// return avatar
+			Ok(Self::Avatar {
 				data_length: avatar.get_length().try_into().unwrap(),
 				name: avatar.take_name(),
 				data: None,
-			});
-		};
-
-		if frame.has_sticker() {
-			fields_count += 1;
+			})
+		} else if frame.has_sticker() {
 			let sticker = frame.take_sticker();
-			ret = Some(Self::Sticker {
+
+			// return sticker
+			Ok(Self::Sticker {
 				data_length: sticker.get_length().try_into().unwrap(),
 				row: sticker.get_rowId(),
 				data: None,
-			});
-		};
-
-		if frame.has_keyValue() {
-			fields_count += 1;
-			let key_value = frame.take_keyValue();
-			ret = Some(Self::KeyValue {
-				key_value
-			});
-		};
-
-		if fields_count != 1 {
-			panic!(
-				"Frame with an unsupported number of fields found, please report to author: {:?}",
+			})
+		} else if frame.has_keyValue() {
+			let mut keyvalue = frame.take_keyValue();
+			let value = if keyvalue.has_blobValue() {
+				KeyValueContent::Blob(keyvalue.take_blobValue())
+			} else if keyvalue.has_booleanValue() {
+				KeyValueContent::Bool(keyvalue.get_booleanValue())
+			} else if keyvalue.has_floatValue() {
+				KeyValueContent::Float(keyvalue.get_floatValue())
+			} else if keyvalue.has_integerValue() {
+				KeyValueContent::Int(keyvalue.get_integerValue().into())
+			} else if keyvalue.has_longValue() {
+				KeyValueContent::Int(keyvalue.get_longValue())
+			} else if keyvalue.has_stringValue() {
+				KeyValueContent::String(keyvalue.take_stringValue())
+			} else {
+				unreachable!()
+			};
+
+			// return keyvalue
+			Ok(Self::KeyValue {
+				key: keyvalue.take_key(),
+				value,
+			})
+		} else {
+			Err(anyhow!(
```

Comment:
Will this ever happen? From what I understand of protobuf and the bindings, a frame can only have fields that are defined in the `Backups.proto`-file. Therefor we should see an unknown frame type. Instead we might get crashes when parsing frames, for example in the `frame.take_...`-calls. I would prefer to keep the old solution of counting fields, which should flag up new fields as well.

Reply:
I'm really not that familiar with protobuf (so I might miss here something) but I think that actually nothing has changed (except that it got a bit easier to understand).

Before:
- All known frame types were checked and if found `frame_count` is increased by one.
- If there is a unknown type, there is no check for it. Thus, `frame_count` is not increased and stays at 0.

After:
- We check all known frame types but end checking if we have found a known frame type. I think, this is valid approach because there won't be a frame with two types (at least this way I understand protobuf).
- If there is a unknown frame type, we don't have a check for it. This way, we enter the else clause and return an error.

If have tested this behavior locally. I removed the preference frame type in if/else clauses and the error is thrown.

Reply:
The Problem is: this only works if we forget to check for a frame type that we know in the protobuf. The protobuf reader will never produce frame types that are not defined in the `Backups.proto` file, i.e. it will never give us an unknown frame type as long as we don't change the Backups.proto file. Howevery "our" Backups.proto is not automatically synced with the version in https://github.com/signalapp/signal-android repo.

When the protobuf parser runs on a backup file that was produced with a different Backups.proto version that contains different frame types, I sometimes gets confused and tells us that there is more than one field in this frame. This is not really an intentional check for new frames, but seems to be a good indicator (if this breaks, the probably is a new frame type).

I hope this clears things up a bit?

Reply:
Indeed, I have now reverted the previous behavior. Thanks!